### PR TITLE
Simplify scene

### PR DESCRIPTION
**Pull request by WojciechBarczynski**
This PR is the continuation of #95 , but more focused on improving the proposed API, rather than writing a new one from scratch. IMO it improves a lot of things in Scene API, simplifies it, and makes in possible to implement, without dealing with the ambiguities introduced with canvas objects.

Original Scene API seems not to follow the Membrane motto `Make simple things easy and hard things possible`, at least in my opinion. This is my take on simplifying it.

Here is some key API info:
- All VC processing is defined by the `Scene` struct
- There are two types of `Objects` (renderable things): `Textures` and `Layouts`
- `Texture` objects have a single input and a single output
- `Layout` objects have multiple inputs and a single output
- `Objects` take input pad name or other `Objects` as inputs
- All `Objects` can be imagined as things taking the simple image(s), processing them, and returning new image
- User of VC can create custom `Transformations` (applied to textures) and `Layouts` (hopefully writing only a little of Elixir, Rust, and shader code)

I would love to hear feedback, to spot problematic things ASAP in the VideoCompositor development process.

### CODE REVIEW

**Review by jerzywilczek — CHANGES_REQUESTED** [review]

File: lib/membrane/video_compositor/scene/layout.ex [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  @layout_enforce_keys [:inputs, :resolution]
+
+  @doc """
+  Returns all required fields in Layout structs.
+  """
+  @spec get_layout_enforce_keys() :: list(enforced_keys_t())
+  def get_layout_enforce_keys() do
+    @layout_enforce_keys
+  end
```
[the remark this answers is not in the document]
Comment (WojciechBarczynski) [reply lib/membrane/video_compositor/scene/layout.ex:44]: I like this idea overall, but it's hard to tell from this spec that `Layout` should be defined by `struct` and `inputs`, `resolution`, and optional fields are fields in that `struct`. What do you think about:
```elixir
@type definition_t :: struct()
@type t :: %{
  :inputs => inputs_t(),
  :resolution => resolution_t(),
  optional(any() => any())
}
```
?

File: lib/membrane/video_compositor/examples/easy.ex [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  while the rest are scaled-down and put in a side strip on top of the main video.
+  """
+
+  alias Membrane.VideoCompositor.Examples.Mock.Layouts.{Overlay, Position}
+  alias Membrane.VideoCompositor.Examples.Mock.Transformations.CornersRounding
+  alias Membrane.VideoCompositor.Scene
+  alias Membrane.VideoCompositor.Scene.{Resolution, Texture}
+
+  @one_third 1.0 / 3.0
+  @sixteen_tenths 16.0 / 9.0
```
[the remark this answers is not in the document]
Comment (WojciechBarczynski) [reply lib/membrane/video_compositor/examples/easy.ex]: Fixed :)

File: lib/membrane/video_compositor/scene/layout.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (jerzywilczek) [reply lib/membrane/video_compositor/scene/layout.ex:44]: I don't really get how this is supposed to be used, you have to elaborate

[the remark this answers is not in the document]
Comment (jerzywilczek) [reply lib/membrane/video_compositor/scene/layout.ex:44]: What do users have to do with this solution to have dialyzer typecheck their own transformations?

File: lib/membrane/video_compositor/examples/hard.ex [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  #   },
+  #   rendering: [
+  #     link(:video_1) |> with_transforms([:rotate]) |> via_in(0) |> to(:three_vids_grid),
+  #     link(:video_1) |> via_in(0) |> to(:merging),
+  #     link(:video_2) |> via_in(1) |> to(:merging),
+  #     link(:merging) |> via_in(1) |> to(:three_vids_grid)
+  #     link(:video_3) |> with_transforms([:to_ball]) |> via_in(2) |> to(:three_vids_grid)
+  #     link(:three_vids_grid) |> to(:output)
+  #   ]
+  # }
```
[the remark this answers is not in the document]
Comment (WojciechBarczynski) [reply lib/membrane/video_compositor/examples/hard.ex:70]: https://membraneframework.atlassian.net/browse/MS-431

File: lib/membrane/video_compositor/scene/layout.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (DominikWolek) [reply lib/membrane/video_compositor/scene/layout.ex:44]: I think we should change how we deal with inputs to the version Bartek proposed, but I'd rather do that in the main PR

**Review by DominikWolek — APPROVED** [review]
I've got no concerns beside other converstions

File: lib/membrane/video_compositor/scene/layout.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (jerzywilczek) [reply lib/membrane/video_compositor/scene/layout.ex:51]: Now that you've implemented this the way I proposed, you should remove the `get_layout_enforce_keys/0` function

File: lib/membrane/video_compositor/examples/mock/layouts/overlay.ex [hunk 4]
```diff
@@ -0,0 +1,24 @@
+defmodule Membrane.VideoCompositor.Examples.Mock.Layouts.Overlay do
+  @moduledoc false
+
+  alias Membrane.VideoCompositor.Examples.Mock.Layouts.Position
+  alias Membrane.VideoCompositor.Scene.{Layout, Object}
+
+  @enforce_keys [:overlay_spec] ++ Layout.get_layout_enforce_keys()
```
Comment (jerzywilczek) [inline lib/membrane/video_compositor/examples/mock/layouts/overlay.ex:7]: `Layout.get_layout_enforce_keys()` should also be removed from here.

File: lib/membrane/video_compositor/examples/mock/layouts/merging.ex [hunk 5]
```diff
@@ -0,0 +1,18 @@
+defmodule Membrane.VideoCompositor.Examples.Mock.Layouts.Merging do
+  @moduledoc """
+  Mock simple component combing two videos / frames / inputs into one.
+  """
+
+  alias Membrane.VideoCompositor.Scene.{Layout, Object, Resolution}
+
+  @enforce_keys Layout.get_layout_enforce_keys()
```
Comment (jerzywilczek) [inline lib/membrane/video_compositor/examples/mock/layouts/merging.ex:8]: `Layout.get_layout_enforce_keys()` should also be removed from here.

File: lib/membrane/video_compositor/examples/mock/layouts/grid.ex [hunk 6]
```diff
@@ -0,0 +1,20 @@
+defmodule Membrane.VideoCompositor.Examples.Mock.Layouts.Grid do
+  @moduledoc """
+  Mock Grid Layout. Videos placements are defined by the :video_count
+  parameter. Mock some components defining different
+  video arrangements for a different number of input sources.
+  """
+
+  alias Membrane.VideoCompositor.Scene.{Layout, Object, Resolution}
+
+  @enforce_keys [:videos_count] ++ Layout.get_layout_enforce_keys()
```
Comment (jerzywilczek) [inline lib/membrane/video_compositor/examples/mock/layouts/grid.ex:10]: `Layout.get_layout_enforce_keys()` should also be removed from here.

**Review by jerzywilczek — CHANGES_REQUESTED** [review]
:cowboy_hat_face:

**Pull request merged**